java-checks/src/main/java/org/sonar/java/checks/SwitchStatementNestedCheck.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.CaseGroupTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+@Rule(key = "S1821")
+public class SwitchStatementNestedCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.SWITCH_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
+    List<CaseGroupTree> groups = switchStatementTree.cases();
+    List<StatementTree> nestedSwitchStatements = groups.stream().flatMap(x -> x.body().stream().filter(t -> t.is(Tree.Kind.SWITCH_STATEMENT))).collect(Collectors.toList());
```

Comment:
This will only catch the `switch` statement which are at NOT nested in other structures, and only detect the ones which are in direct child of the `body()`. For instance, the following code won't be detected while **it should**:

```
switch(myVar) {
  case A: {
    switch(myOtherVar) { // switch nested in the block
      case STUFF:
        return 42;
      default:
        break;
    }
  }
  default:
    break;
}
```

In order to catch these, you will need to use a child `BaseTreeVisitor` to look for `switch` within the body.
Feel free to ask me how to do it.